File: src/api/java/Statistics.java
/**
Copyright (c) 2012-2014 Microsoft Corporation
   
Module Name:

    Statistics.java

Abstract:

Author:

    @author Christoph Wintersteiger (cwinter) 2012-03-15

Notes:
    
**/ 

package com.microsoft.z3;

/**
 * Objects of this class track statistical information about solvers.
 **/
public class Statistics extends Z3Object {
    /**
     * Statistical data is organized into pairs of [Key, Entry], where every
     * Entry is either a {@code DoubleEntry} or a {@code UIntEntry}
     **/
    public class Entry
    {
        /**
         * The key of the entry.
         **/
        public String Key;

        /**
         * The uint-value of the entry.
         **/
        public int getUIntValue()
        {
            return m_int;
        }

        /**
         * The double-value of the entry.
         **/
        public double getDoubleValue()
        {
            return m_double;
        }

        /**
         * True if the entry is uint-valued.
         **/
        public boolean isUInt()
        {
            return m_is_int;
        }

        /**
         * True if the entry is double-valued.
         **/
        public boolean isDouble()
        {
            return m_is_double;
        }

        /**
         * The string representation of the entry's value.
         * 
         * @throws Z3Exception
         **/
        public String getValueString()
        {
            if (isUInt()) {
                return Integer.toString(m_int);
            } else if (isDouble()) {
                return Double.toString(m_double);
            } else {
                throw new Z3Exception("Unknown statistical entry type");
            }
        }

        /**
         * The string representation of the Entry.
         **/
        @Override
        public String toString() {
            return Key + ": " + getValueString();
        }

        private boolean m_is_int = false;
        private boolean m_is_double = false;
        private int m_int = 0;
        private double m_double = 0.0;

        Entry(String k, int v)
        {
            Key = k;
            m_is_int = true;
            m_int = v;
        }

        Entry(String k, double v)
        {
            Key = k;
            m_is_double = true;
            m_double = v;
        }
    }

    /**
     * A string representation of the statistical data.
     **/
    @Override
    public String toString()
    {
        return Native.statsToString(getContext().nCtx(), getNativeObject());
    }

    /**
     * The number of statistical data.
     **/
    public int size()
    {
        return Native.statsSize(getContext().nCtx(), getNativeObject());
    }

    /**
     * The data entries.
     * 
     * @throws Z3Exception
     **/
    public Entry[] getEntries()
    {

        int n = size();
        Entry[] res = new Entry[n];
        for (int i = 0; i < n; i++)
        {
            Entry e;
            String k = Native.statsGetKey(getContext().nCtx(), getNativeObject(), i);
            if (Native.statsIsUint(getContext().nCtx(), getNativeObject(), i)) {
                e = new Entry(k, Native.statsGetUintValue(getContext().nCtx(),
                    getNativeObject(), i));
            } else if (Native.statsIsDouble(getContext().nCtx(), getNativeObject(), i)) {
                e = new Entry(k, Native.statsGetDoubleValue(getContext().nCtx(),
                    getNativeObject(), i));
            } else {
                throw new Z3Exception("Unknown data entry value");
            }
            res[i] = e;
        }
        return res;
    }

    /**
     * The statistical counters.
     **/
    public String[] getKeys()
    {
        int n = size();
        String[] res = new String[n];
        for (int i = 0; i < n; i++)
            res[i] = Native.statsGetKey(getContext().nCtx(), getNativeObject(), i);
        return res;
    }

    /**
     * The value of a particular statistical counter.
     * Remarks: Returns null if
     * the key is unknown.
     * 
     * @throws Z3Exception
     **/
    public Entry get(String key)
    {
        int n = size();
        Entry[] es = getEntries();
        for (int i = 0; i < n; i++) {
            if (es[i].Key.equals(key)) {
                return es[i];
            }
        }
        return null;
    }

    Statistics(Context ctx, long obj)
    {
        super(ctx, obj);
    }

    @Override
    void incRef() {
        getContext().getStatisticsDRQ().storeReference(getContext(), this);
    }

    @Override
    void addToReferenceQueue() {
        Native.statsIncRef(getContext().nCtx(), getNativeObject());
    }
}

File: src/api/dotnet/Statistics.cs
/*++
Copyright (c) 2012 Microsoft Corporation

Module Name:

    Statistics.cs

Abstract:

    Z3 Managed API: Statistics

Author:

    Christoph Wintersteiger (cwinter) 2012-03-22

Notes:
    
--*/

using System;
using System.Diagnostics;
namespace Microsoft.Z3
{

    using Z3_context = System.IntPtr;
    using Z3_stats = System.IntPtr;
    /// <summary>
    /// Objects of this class track statistical information about solvers. 
    /// </summary>
    public class Statistics : Z3Object
    {
        /// <summary>
        /// Statistical data is organized into pairs of [Key, Entry], where every
        /// Entry is either a <c>DoubleEntry</c> or a <c>UIntEntry</c>
        /// </summary>
        public class Entry
        {
            /// <summary>
            /// The key of the entry.
            /// </summary>
            readonly public string Key;
            /// <summary>
            /// The uint-value of the entry.
            /// </summary>
            public uint UIntValue { get { return m_uint; } }
            /// <summary>
            /// The double-value of the entry.
            /// </summary>
            public double DoubleValue { get { return m_double; } }
            /// <summary>
            /// True if the entry is uint-valued.
            /// </summary>
            public bool IsUInt { get { return m_is_uint; } }
            /// <summary>
            /// True if the entry is double-valued.
            /// </summary>
            public bool IsDouble { get { return m_is_double; } }

            /// <summary>
            /// The string representation of the entry's value.
            /// </summary>
            public string Value
            {
                get
                {

                    if (IsUInt)
                        return m_uint.ToString();
                    else if (IsDouble)
                        return m_double.ToString();
                    else
                        throw new Z3Exception("Unknown statistical entry type");
                }
            }

            /// <summary>
            /// The string representation of the Entry.
            /// </summary>
            public override string ToString()
            {
                return Key + ": " + Value;
            }

            #region Internal
            readonly private bool m_is_uint = false;
            readonly private bool m_is_double = false;
            readonly private uint m_uint = 0;
            readonly private double m_double = 0.0;
            internal Entry(string k, uint v)
            {
                Key = k;
                m_is_uint = true;
                m_uint = v;
            }
            internal Entry(string k, double v)
            {
                Key = k;
                m_is_double = true;
                m_double = v;
            }
            #endregion
        }

        /// <summary>
        /// A string representation of the statistical data.
        /// </summary>    
        public override string ToString()
        {
            return Native.Z3_stats_to_string(Context.nCtx, NativeObject);
        }

        /// <summary>
        /// The number of statistical data.
        /// </summary>
        public uint Size
        {
            get { return Native.Z3_stats_size(Context.nCtx, NativeObject); }
        }

        /// <summary>
        /// The data entries.
        /// </summary>
        public Entry[] Entries
        {
            get
            {
                return NativeEntries(Context.nCtx, NativeObject);
            }
        }

        internal static Entry[] NativeEntries(Z3_context ctx, Z3_stats stats)
        {
            uint n = Native.Z3_stats_size(ctx, stats);
            Entry[] res = new Entry[n];
            for (uint i = 0; i < n; i++)
            {
                Entry e;
                string k = Native.Z3_stats_get_key(ctx, stats, i);
                if (Native.Z3_stats_is_uint(ctx, stats, i) != 0)
                    e = new Entry(k, Native.Z3_stats_get_uint_value(ctx, stats, i));
                else if (Native.Z3_stats_is_double(ctx, stats, i) != 0)
                    e = new Entry(k, Native.Z3_stats_get_double_value(ctx, stats, i));
                else
                    throw new Z3Exception("Unknown data entry value");
                res[i] = e;
            }
            return res;
        }

        /// <summary>
        /// The statistical counters.
        /// </summary>
        public string[] Keys
        {
            get
            {

                uint n = Size;
                string[] res = new string[n];
                for (uint i = 0; i < n; i++)
                    res[i] = Native.Z3_stats_get_key(Context.nCtx, NativeObject, i);
                return res;
            }
        }

        /// <summary>
        /// The value of a particular statistical counter.
        /// </summary>        
        /// <remarks>Returns null if the key is unknown.</remarks>
        public Entry this[string key]
        {
            get
            {
                uint n = Size;
                Entry[] es = Entries;
                for (uint i = 0; i < n; i++)
                    if (es[i].Key == key)
                        return es[i];
                return null;
            }
        }

        #region Internal
        internal Statistics(Context ctx, IntPtr obj)
            : base(ctx, obj)
        {
            Debug.Assert(ctx != null);
        }

        internal override void IncRef(IntPtr o)
        {
            Native.Z3_stats_inc_ref(Context.nCtx, o);
        }

        internal override void DecRef(IntPtr o)
        {
            lock (Context)
            {
                if (Context.nCtx != IntPtr.Zero)
                    Native.Z3_stats_dec_ref(Context.nCtx, o);
            }
        }
        #endregion
    }
}
